blocks/graph.py

```diff
@@ -421,6 +423,91 @@ def apply_noise(computation_graph, variables, level, seed=None):
     return computation_graph.replace(replace)
 
 
+def collect_parameters(computation_graph, params):
+    """Replace parameters with a single shared variable.
+
+    This can be useful if you need to calculate the full Hessian of a
+    computational graph. It replaces parameters with slices of a single
+    large vectors like
+
+    >>> from blocks.utils import shared_floatx
+    >>> W1 = shared_floatx(numpy.random.rand(10, 10))
+    >>> W2 = shared_floatx(numpy.random.rand(10, 10))
+    >>> all_params = shared_floatx(numpy.concatenate(
+    ...     [W1.get_value().flatten(), W2.get_value().flatten()]))
+    >>> W1 = all_params[:W1.size]
+    >>> W2 = all_params[W1.size:]
+
+    Parameters
+    ----------
+    computation_graph : :class:`ComputationGraph` instance
+        The managed Theano graph in which to collect parameters.
+    params : list of Theano shared variables
+        The parameters whose values should be collected.
+
+    Returns
+    -------
+    ComputationGraph instance
+        A new Theano graph which has all the given parameters collected
+        into a single large shared variable.
+
+    Notes
+    -----
+    Note that this replacement makes the training of the model
+    significantly slower because of the large amount of Theano's
+    ``set_subtensor`` calls needed to train the model.
+
+    Examples
+    --------
+    >>> from blocks.bricks import MLP, Sigmoid
+    >>> from blocks.bricks.cost import SquaredError
+    >>> from theano import tensor
+    >>> x = tensor.matrix()
+    >>> mlp = MLP(activations=[Sigmoid(), Sigmoid()], dims=[784, 100, 784])
+    >>> cost = SquaredError().apply(x, mlp.apply(x))
+    >>> cg = ComputationGraph(cost)
+    >>> new_cg = collect_parameters(cg, cg.shared_variables)
+
+    The new graph only has a single shared variable. This variable receives
+    the :const:`COLLECTOR` role.
+
+    >>> new_cg.shared_variables
+    [collected_params]
+
+    The bricks' variables have been replaced with reshaped segments of this
+    single shared variable. These replacements are given the
+    :const:`.COLLECTED` role.
+
+    >>> from blocks.filter import VariableFilter
+    >>> from blocks.roles import PARAMETER
+    >>> var_filter = VariableFilter(roles=[COLLECTED])
+    >>> var_filter(new_cg.variables)  # doctest: +SKIP
+    [Reshape{1}.0, Reshape{1}.0, Reshape{2}.0, Reshape{2}.0]
+
+    """
+    param_values, param_sizes, param_shapes = [], [], []
+    for param in params:
+        param_values.append(param.get_value(borrow=True))
+        param_sizes.append(param_values[-1].size)
+        param_shapes.append(param_values[-1].shape)
+
+    new_params = shared_floatx_zeros(sum(param_sizes))
+    new_params.set_value(numpy.concatenate([value.flatten()
+                                            for value in param_values]))
+    new_params.name = 'collected_params'
+    add_role(new_params, COLLECTOR)
```

Comment:
Sorry, forgot about this!

I think it would be perfectly okay to give the collector a `PARAMETER` role. I do not consider it as a role copying, you are introducing something new into the graph and you should be allowed to give it a role.

The algorithm of parameter discovery you propose seems quite custom to me.  I am not sure that we should have something like that as our default logic. For all other graph substitutions we currently have the original shared variables keep being a part of the graph and will be found by `ComputaionGraph.parameters` call.